File: solutions/2026-02-25/javascript/2026-02-25-sort-integers-by-the-number-of-1-bits.js
/**
 * @file Sort Integers by The Number of 1 Bits
 * @link https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/
 *
 * Problem Summary: Sort an array of integers first by the count of 1s in their binary representation (ascending),
 * and then by the integer value itself (ascending) for ties.
 *
 * Approach:
 * We can use a custom sorting function for the array. The custom sorting function will take two numbers, `a` and `b`,
 * and determine their order.
 *
 * 1. Count the number of set bits (1s) in the binary representation of `a` and `b`.
 *    - A common way to count set bits is to repeatedly check the last bit using the bitwise AND operator (`& 1`)
 *      and then right-shift the number (`>>= 1`) until the number becomes 0.
 *    - Alternatively, for positive integers, the `toString(2)` method followed by counting '1's can be used,
 *      but bitwise operations are generally more efficient.
 *
 * 2. Compare the bit counts:
 *    - If `countSetBits(a) < countSetBits(b)`, then `a` should come before `b`. Return -1.
 *    - If `countSetBits(a) > countSetBits(b)`, then `b` should come before `a`. Return 1.
 *
 * 3. If the bit counts are equal, compare the numbers themselves:
 *    - If `a < b`, then `a` should come before `b`. Return -1.
 *    - If `a > b`, then `b` should come before `a`. Return 1.
 *    - If `a == b`, their order doesn't matter. Return 0.
 *
 * We can implement a helper function `countSetBits` to perform the bit counting.
 *
 * Time Complexity:
 * The sorting operation itself typically has a time complexity of O(N log N), where N is the length of the array.
 * For each comparison within the sort, we call `countSetBits`. The `countSetBits` function iterates at most `log(max_val)` times,
 * where `max_val` is the maximum value in the array (since we are dealing with integers up to 10^4, this is a small constant, around 14 bits).
 * Therefore, the overall time complexity is O(N log N * log(max_val)). Given the constraints, log(max_val) is very small, so it's effectively O(N log N).
 *
 * Space Complexity:
 * The space complexity is primarily determined by the sorting algorithm used by the JavaScript engine. Most in-place sorting algorithms
 * have a space complexity of O(log N) or O(1) (for the call stack or auxiliary space). The `countSetBits` function uses O(1) auxiliary space.
 * Thus, the space complexity is O(log N) or O(1) depending on the sort implementation.
 */

/**
 * Counts the number of set bits (1s) in the binary representation of a non-negative integer.
 * @param {number} n The input integer.
 * @returns {number} The count of set bits.
 */
const countSetBits = (n) => {
    let count = 0;
    // Iterate while n is greater than 0.
    while (n > 0) {
        // Check if the last bit is 1 using bitwise AND with 1.
        // If n & 1 is 1, it means the last bit is 1.
        count += (n & 1);
        // Right-shift n by 1 bit to process the next bit.
        // This is equivalent to integer division by 2.
        n >>= 1;
    }
    return count;
};

/**
 * @param {number[]} arr
 * @return {number[]}
 */
const sortByBits = function(arr) {
    // Use the built-in sort method with a custom comparison function.
    arr.sort((a, b) => {
        // Count set bits for both numbers.
        const bitsA = countSetBits(a);
        const bitsB = countSetBits(b);

        // First, compare by the number of set bits.
        if (bitsA !== bitsB) {
            // If bitsA is less than bitsB, 'a' comes first.
            // If bitsA is greater than bitsB, 'b' comes first.
            return bitsA - bitsB;
        } else {
            // If the number of set bits is the same, sort by the integer value itself.
            // If 'a' is less than 'b', 'a' comes first.
            // If 'a' is greater than 'b', 'b' comes first.
            return a - b;
        }
    });

    // Return the sorted array.
    return arr;
};
```

File: solutions/2026-02-25/java/2026-02-25-sort-integers-by-the-number-of-1-bits.java
// Problem: Sort Integers by The Number of 1 Bits
// Link: https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/
//
// Approach:
// The problem requires sorting an array of integers based on two criteria:
// 1. The number of set bits (1s) in their binary representation (ascending).
// 2. If the number of set bits is the same, then sort by the integer value itself (ascending).
//
// We can achieve this by using a custom comparator with Java's `Arrays.sort()` method.
// The comparator will first count the set bits for each pair of numbers.
// If the counts are different, it sorts based on the counts.
// If the counts are the same, it sorts based on the actual integer values.
//
// To count the set bits in an integer, we can use the `Integer.bitCount()` method provided by Java,
// which efficiently counts the number of set bits.
//
// Time Complexity:
// The `Arrays.sort()` method for primitive arrays typically uses a dual-pivot quicksort algorithm,
// which has an average time complexity of O(N log N), where N is the length of the array.
// The custom comparator involves `Integer.bitCount()`, which takes O(log M) time where M is the maximum
// value of an integer (or O(1) as the number of bits in an integer is fixed, e.g., 32 for int).
// Therefore, the overall time complexity is O(N log N * log M) or effectively O(N log N) because
// `Integer.bitCount()` is very efficient.
//
// Space Complexity:
// The `Arrays.sort()` method, in its in-place implementation (or with limited auxiliary space),
// typically has a space complexity of O(log N) for the recursion stack in quicksort, or O(1)
// for some iterative sorts. The `Integer.bitCount()` method uses constant extra space.
// Thus, the overall space complexity is O(log N) or O(1) depending on the specific sort implementation details.
// For Java's `Arrays.sort` on primitive arrays, it's generally considered O(log N) due to recursion.
//
class Solution {
    public int[] sortByBits(int[] arr) {
        // Use Arrays.sort with a custom comparator.
        // The comparator defines the sorting logic.
        // We need to sort the array of primitive integers, so we'll box them into an Integer array
        // to use the comparator, and then convert back to an int array.
        Integer[] boxedArr = new Integer[arr.length];
        for (int i = 0; i < arr.length; i++) {
            boxedArr[i] = arr[i];
        }

        // Sort the boxed array using a lambda expression as a custom comparator.
        java.util.Arrays.sort(boxedArr, (a, b) -> {
            // Count the number of set bits (1s) in the binary representation of 'a' and 'b'.
            int bitsA = Integer.bitCount(a);
            int bitsB = Integer.bitCount(b);

            // First, compare based on the number of set bits.
            if (bitsA != bitsB) {
                // If the number of bits is different, sort in ascending order of bit count.
                return bitsA - bitsB;
            } else {
                // If the number of bits is the same, sort in ascending order of the integer values.
                return a - b;
            }
        });

        // Convert the sorted boxed array back to a primitive int array.
        int[] sortedArr = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            sortedArr[i] = boxedArr[i];
        }

        // Return the sorted array.
        return sortedArr;
    }
}
